xblock/query.py

```diff
@@ -0,0 +1,111 @@
+"""
+    Define Query and Shared objects. Query supports ad-hoc queries via get, find, set operators.
+    Shared uses `bind` functions to access data from pre-defined entities.  
+"""
+
+from copy import deepcopy
+
+class Query(object):
+    """
+    Class for handling remote query operations
+    """
+
+    def __init__(self, field_name, remote_scope, bind):
+        self._queryable = Queryable()
+        self._queryable.bind(field_name, remote_scope, bind)
+
+    def __get__(self, field, field_class):
+        """Get the queryable instance
+        
+        Args:
+            xblock (TYPE): Description
+            xblock_class (TYPE): Description
+        
+        Returns:
+            TYPE: Description
+        """
+
+        return self._queryable
+
+    def __set__(self, field, val):
+        print val
+
+class Shared(object):
+    """
+    Class for handling shared field operations
+    """
+    def __init__(self):
+        pass
+
+
+class Queryable(object):
+    """
+    Class for Queryable objects
+    """
+
+    def __init__(self):
+        self._field_name = "unknown"
+        self._remote_scope = None
+        self._bind = None
+
+    @property
+    def field_name(self):
+        return self._field_name
+
+    def bind(self, field_name, remote_scope, bind):
+        """Summary
+        
+        Args:
+            field_name (TYPE): Description
+            remote_scope (TYPE): Description
+            bind (TYPE): Description
+        
+        Returns:
+            TYPE: Description
+        """
+        self._field_name = field_name
+        self._remote_scope = remote_scope
+        self._bind = bind
+
+    def _replace_xblock_user_id(self, xblock, user_id):
+        new_block = deepcopy(xblock)
```

Comment:
What is XBlock right here? Is it the actual XBlock object? If so, could a deepcopy like this may have unintended side effects? I'm not familiar enough with this code to know, but I'd like to flag for @cpennington to have a look.